### PR TITLE
chore: event bus usage pr5

### PR DESCRIPTION
**Pull request by Shahroz16**
Broken down and reworked PR for the Event Bus implementation.

More Details about the previous implementation and flow can be found in the previous [PR](https://github.com/customerio/customerio-ios/pull/440)

PR 5: Holds the usage of the event bus handler and its working.

### CODE REVIEW

File: Sources/DataPipeline/DataPipelineImplementation.swift
Comment (levibostian) [inline Sources/DataPipeline/DataPipelineImplementation.swift:46]: I don't find comments like this valuable in the code. From reading the lines of code, I can see that a plugin is being added and journeys events are being subscribed to. 

If a comment is added, I would find them valuable if the comments mention why we are adding a plugin and why we are subscribing to journeys events.

File: Sources/DataPipeline/plugins/DataPipelinePublishedEvents.swift
Comment (levibostian) [inline Sources/DataPipeline/plugins/DataPipelinePublishedEvents.swift:8]: I do not understand why this class is needed? 

From my knowledge of segment plugins, here is an example flow for how I believe this plugin would work: 

1. A profile is identified in the Segment SDK. 
2. Segment SDK calls this plugin's `identify(event: IdentifyEvent)` function
3. EventBus posts new `ProfileIdentifiedEvent` event. <--- why do this need to happen? 

My biggest question is: Why do we need our EventBus to post a new `ProfileIdentifiedEvent` event?

I was expecting that once a profile is identified in the Segment SDK, our EventBus should have already posted an `ProfileIdentifiedEvent` event. 

I was expecting a flow like this: 

1. Profile identified in the CIO SDK. 
2. From the CIO SDK, EventBus posts new `ProfileIdentifiedEvent` event. 
3. A profile is identified in the Segment SDK. 

From the flow above, I don't see a need for this plugin. 

---

If there is a use case that I am not aware of that this plugin solves, I would appreciate editing the comment in this class explaining why it's needed.

**Review by levibostian — CHANGES_REQUESTED** [review]
PR looks good. I do have 1 blocker: https://github.com/customerio/customerio-ios/pull/453/files#r1440665476 that I want to better understand before approval.

File: Sources/DataPipeline/plugins/DataPipelinePublishedEvents.swift
Comment (Shahroz16) [reply Sources/DataPipeline/plugins/DataPipelinePublishedEvents.swift:8]: I am not sure if i fully understand the question, but I would like to explain the flow to see if it helps resolve any confusion.

Before I respond, I would like to point to one aspect of plugins in the Segment. 

> Plugins are run through a timeline, which executes in order of insertion based on their types

So the very first line, explain at what point this plugin would execute  
`var type: Segment.PluginType = .before`

> Why do we need our EventBus to post a new ProfileIdentifiedEvent event?

I tried to answer this with this comment

>/// This plugin allows for decoupled modules to subscribe and react to these events.

When a profile is identified, we want to let `InApp` module know what `user id` to set, similarly screen, and the same goes for reset. Similarly, Push can subscribe to its events. 

>I do not understand why this class is needed.

I tried to answer that with
> /// This class serves as a centralized hub for dispatching events generated from DataPipeline module.
/// It is designed to reduce redundancy and enhance maintainability by funneling these operations through a single plugin.

We have multiple `identify` methods, some accept `identifier` some `user id`, similarly some would accept, `map` or `codable`. Similarly, we would have multiple screen methods and the same for others. 

Instead of posting in the `event bus` from all of those methods or trying to create a common method you then call from all of those methods. This plugin offers a centralised way of handling it. 

No matter if you call,
`CustomerIO.identify(idenfity) or CustomerIO.identify(userId)` etc, it will end up at this plugin which will then publish that event for any other module that might be subscribing to it.

File: Sources/Common/autogenerated/AutoDependencyInjection.generated.swift
Comment (Ahmed-CIO) [inline Sources/Common/autogenerated/AutoDependencyInjection.generated.swift:519]: I was going to comment on this part and totally forgot it is autogenerated -_-

File: Sources/DataPipeline/plugins/DataPipelinePublishedEvents.swift
Comment (Ahmed-CIO) [reply Sources/DataPipeline/plugins/DataPipelinePublishedEvents.swift:8]: looking at the code and this conversation, I believe this conversation doesn't immediately result in anything actionable in the code itself. Maybe a chat followed by a conclusion or a documentation can help instead of having this conversation over the PR and going through longer feedback loop

File: Sources/DataPipeline/DataPipelineImplementation.swift
Comment (Ahmed-CIO) [reply Sources/DataPipeline/DataPipelineImplementation.swift:46]: +1, please clean up the comments before merging

**Review by Ahmed-CIO — APPROVED** [review]
I am unblocking this one on the basis that:
- The code itself looks good to go.
- The open questions can be addressed without blocking the PR

If addressing the questions resulted in a need to change anything, a follow up PR will be published to address these comments

**Pull request merged**
